### DISCUSSION TITLE
(not) using additional dependencies for a device

### DISCUSSION THREAD

**Discussion by bklebel** [pymeasure/pymeasure#1149] [Q&A] [answered]
I am about to use a thermocouple readout chip, specifically the microchip mcp9600. For convenience, and interoperability I wanted to make a wrapping pymeasure Instrument, around a pre-existing library, specifically [this one](https://github.com/pimoroni/mcp9600-python) ([pypi](https://pypi.org/project/mcp9600/)). 

It would be nice to actually use the library, to give appropriate credit and such, however, I do not want to ask this specific library to become a dependency for pymeasure. 
Any ideas about how I could best go about that? 

An alternative would be to just "fork" it (MIT license) and include it in pymeasure in full. However, then I am at the next problem, since it again uses another library, i2cdevice. For now, I guess I will work with a local version where I use the library as is, but once I have an instrument ready, I would like to share it. Also, I would like to avoid the mess of a local pymeasure installation with my own changes, and keeping it up to date. 
It would be possible of course, to write a new i2c adapter, and port the whole library into pymeasure, keeping it purely pymeasure. I am not sure I'll have the time to do that (right now).

**Comment by BenediktBurger** [comment, marked as answer]
Sounds like a case for #1058 . Even though it does not require a dll or such, it is more than using current pymeasure dependencies.
In that issue we were thinking about additional repositories for dll driven drivers which could be created as pymeasure plugins.

Maybe that could work for you as well (we did not continue creating a plugin mechanics yet).

**Reply by bklebel** [reply]
thanks, that sounds like a good idea. In the meantime, I did consider writing an i2c adapter more carefully, since this is a fairly common protocol for microelectronics, and people who need/want to do more home-made custom setups than expensive equipment might then benefit from a reasonably easy way into pymeasure too. 

Do you have any ideas on how soon the plugin mechanics might be available?

**Reply by BenediktBurger** [reply]
I think nobody is currently working on it due to time constraints.
The best approach, in my opinion, would be an independent repository (maybe linking there from a pymeasure announcement discussion).
Then later we can decide whether to include the code in pymeasure itself or make a pymeasure plugin out of it.

Any other ideas @bilderbuchi ?

**Reply by bilderbuchi** [reply]
It sounds to me like the namespace package/plugin approach sketched in #1058 sounds like the best approach for your problem. Nobody has been working on that, unfortunately. 
I have no other/better ideas :-| 

But yeah, let's keep your work external for now, develop, and as you are close to the project anyway, @bklebel, I think any potential refactoring later on should not be a problem.
If I understand the situation correctly, it seems like a separate repo/package where you create/maintain an `Instrument` for your chip (potentially an `Adapter`, too?), and which depends on both `pymeasure` and `mcp9600`, should cover your usecase and avoids blokcing you on pymeasure progress?

**Reply by bklebel** [reply]
Thanks for your input! 
Yes, I guess just an `Instrument` in my project repo will be enough for now. Once a plugin approach becomes available, I can implement such a plugin for this device. 

For now, I don't use an `Adapter`, but just do some monkey patching of the `Instrument` class without `super.__init__()`, and handle the comms with the chip manually. It's now in a first rudimentary working stage; let's see whether I can find the time to clean it up into an i2c-adapter later (will depend on how much time my current employer will allow for open-source maintenance during working hours).
